### PR TITLE
feat: support custom formvalue function

### PR DESCRIPTION
**Pull request by li-jin-gou**
Some gin or net/http services have a different formvalue  function when migrating to fasthttp.
it maybe causes some bugs.

net/http:
```
// FormValue returns the first value for the named component of the query.
// POST and PUT body parameters take precedence over URL query string values.
// FormValue calls ParseMultipartForm and ParseForm if necessary and ignores
// any errors returned by these functions.
// If key is not present, FormValue returns the empty string.
// To access multiple values of the same key, call ParseForm and
// then inspect Request.Form directly.
```

The user can customize a formvalue function to avoid this problem, but it would be better if the framework provided this switch.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -1108,6 +1108,12 @@ func SaveMultipartFile(fh *multipart.FileHeader, path string) (err error) {
 //
 // The returned value is valid until your request handler returns.
 func (ctx *RequestCtx) FormValue(key string) []byte {
+	return defaultFormValue(ctx, key)
+}
+
+type FormValue func(*RequestCtx, string) []byte
```
Comment (erikdubbelboer) [inline server.go:1114]:
```suggestion
type FormValueFunc func(*RequestCtx, string) []byte
```

File: server.go [hunk 2]
```diff
@@ -1126,6 +1132,34 @@ func (ctx *RequestCtx) FormValue(key string) []byte {
...
 }
 
+// SetCustomFormValueFunc sets FormValue function implementation to get form value.
+func SetCustomFormValueFunc(f FormValue) {
+	defaultFormValue = f
+}
+
+// SetStandardFormValueFunc sets FormValue function implementation to get form value.
+// Consistent behaviour with net/http. POST and PUT body parameters take precedence over URL query string values.
+func SetStandardFormValueFunc() {
```
Comment (erikdubbelboer) [inline server.go:1142]:
```suggestion
func SetNetHttpFormValueFunc() {
```

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]
Do you have a real world example of this going wrong or is it something you think might go wrong?

File: server.go (hunk 1 above)
Comment (li-jin-gou) [reply server.go]: done

File: server.go (hunk 2 above)
Comment (li-jin-gou) [reply server.go]: done

**Comment by li-jin-gou** [comment]
> Do you have a real world example of this going wrong or is it something you think might go wrong?

There was a problem with the code in the real environment and I just noticed the difference. the front end passes the same parameters and different values in the http body and query and result in an unintended bug...😩

File: server.go [hunk 3]
```diff
@@ -412,6 +412,7 @@ type Server struct {
 	concurrencyCh    chan struct{}
 	perIPConnCounter perIPConnCounter
 	serverName       atomic.Value
+	FormValueFunc    FormValueFunc
```
Comment (erikdubbelboer) [inline server.go:415]: This field needs some documentation since it's exposed.

File: server.go [hunk 4]
```diff
@@ -1108,23 +1110,54 @@ func SaveMultipartFile(fh *multipart.FileHeader, path string) (err error) {
...
+		}
+		return nil
 	}
-	mf, err := ctx.MultipartForm()
-	if err == nil && mf.Value != nil {
-		vv := mf.Value[key]
-		if len(vv) > 0 {
-			return []byte(vv[0])
+
+	// NetHttpFormValueFunc Consistent behaviour with net/http. POST and PUT body parameters take precedence over URL query string values.
```
Comment (erikdubbelboer) [inline server.go:1141]:
```suggestion
	// NetHttpFormValueFunc gives consistent behavior with net/http. POST and PUT body parameters take precedence over URL query string values.
```

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]

File: server.go (hunk 4 above)
Comment (li-jin-gou) [reply server.go:1141]: thanks

File: server.go (hunk 3 above)
Comment (li-jin-gou) [reply server.go]: thanks and done

**Comment by erikdubbelboer** [comment]
Thanks!

**Pull request merged**